Basic "Run" command

brendandburns opened the pull request:
@bgrant0607 @smarterclayton @ghodss 

I tried to address the concerns in https://github.com/GoogleCloudPlatform/kubernetes/issues/1695.

Concretely:
- Generator is a separate library
- I don't think generator versioning is necessary, we generate using internal representation, than convert to versioned output, so output is versioned, but generator is pegged to internal representation
- Creates a single replication controller, which is printed when things are done, I argue that object is easy to discover.
- There is a `--preview` flag that is suitable for "just printing" and all resource printers are supported.
- No special impertative commands, just generate, and then optionally send on wire, then print.
- Exposed some functionality, open to further suggestions.

Please let me know thoughts.  I'll add tests when folks agree this is a good starting place.

Thanks!
--brendan

brendandburns committed:
Add an initial run command

bgrant0607 commented:
Thanks. I'll take a look.

On generator versioning: Our internal equivalent of generators became essentially unchangeable due to the lack of versioning. We don't need to surface it through the command line now, but we are going to need it at some point. This is not related to versioning of the API itself -- the generator should be able to leverage our conversion logic.

bgrant0607 commented:
Anyone who looks at this should also read the discussion on #3233, in addition to #1695.

bgrant0607 commented on pkg/kubectl/cmd/run.go:
```diff
@@ -0,0 +1,111 @@
+  $ kubectl run nginx dockerfile/nginx
+  <starts a single instance of nginx>
+
+  $ kubectl run nginx dockerfile/nginx --command="nginx run" --count=5
```
This syntax is entirely unacceptable, for reasons which I've discussed thoroughly in #1695 and #3233. Will defer further review pending IRL discussion.

bgrant0607 replied:
Summary from discussion:
* `run` will be changed to something that makes it clear that it's only one container: `runcontainer`, `runone`, `runimage`, ...
* `--preview` will be changed to `--dryrun` and made available to other commands (though support in those commands wouldn't be implemented in this PR). Example proposed behavior: get --dryrun could list URLs, create --dryrun could dump json, update --dryrun could dump json and report conflicts, delete --dryrun could list URLs.
* Command-line flags will be mechanically convertible to json API field names.
* Generator will be versioned.
* There will be an option for embedded podspec json on the command line.

bgrant0607 replied:
When we add a controller for RestartPolicyOnFailure pods, this command should generate the appropriate controller.

brendandburns commented:
Ok, @bgrant0607 and I hashed this out IRL.
To address things:
  * Refactored this out into a generator interface and a syntactic sugar that calls into that, added a version to the template.
  * Changed ```--preview``` to ```--dry-run```
  * Made sure that all flags match the underlying API object field names
  * Renamed ```run``` to ```run-container```

We discussed adding in inline JSON overrides, but it's actually more complicated than I initially thought, I will address in a different PR.

bgrant0607 commented on docs/cli.md:
```diff
@@ -125,6 +125,63 @@ Usage:
+  $ kubectl run nginx dockerfile/nginx --command="nginx run" --count=5
+  <starts a replicated instance of nginx, with the specified command for the container>
+  
+  $ kubectl run nginx dockerfile/nginx --preview
```
Should be --dry-run now.

bgrant0607 replied:
More generally, please update this doc to the new syntax.

bgrant0607 commented:
Thanks. Looking. It's ok to punt on json for now.

bgrant0607 commented on pkg/kubectl/cmd/helpers.go:
```diff
@@ -171,3 +173,17 @@ func ReadConfigDataFromLocation(location string) ([]byte, error) {
 	}
 }
+
+func PrintObject(cmd *cobra.Command, obj runtime.Object, f *Factory, out io.Writer) {
```
What's the reason this isn't in printing.go?

brendandburns replied:
Done.

satnam6502 commented on pkg/kubectl/run.go:
```diff
@@ -0,0 +1,100 @@
+		labels = ParseLabels(labelString)
+	} else {
+		labels = map[string]string{
+			"name": params["name"],
```
How about
```labels = labels.SelectorFromSet(labels.Set{"name": name}.AsSelector())
```
instead of:
```
labels = map[string]string{
	     "name": params["name"]}
```

brendandburns replied:
Selector interface won't work when I try to put it back into the api object (```Selector``` is not ```map[string]string```)

satnam6502 replied:
I wrote the wrong thing at first, I updated it to: labels.Set{"name":
name}.AsSelector()

S

On Tue, Jan 13, 2015 at 1:45 PM, Brendan Burns <notifications@github.com>
wrote:

> In pkg/kubectl/run.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3424#discussion-diff-22897598>
> :
>
> > +		{"labels", false},
> > +		{"name", true},
> > +		{"replicas", true},
> > +		{"image", true},
> > +		{"command", false},
> > +	}
> > +}
> > +
> > +func (BasicReplicationController) Generate(params map[string]string) (runtime.Object, error) {
> > +	labelString, found := params["labels"]
> > +	var labels map[string]string
> > +	if found && len(labelString) > 0 {
> > +		labels = ParseLabels(labelString)
> > +	} else {
> > +		labels = map[string]string{
> > +			"name": params["name"],
>
> Selector interface won't work when I try to put it back into the api
> object (Selector is not map[string]string)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3424/files#r22897598>
> .
>

satnam6502 replied:
Ah yes, right, the types might not work out then.

On Tue, Jan 13, 2015 at 1:47 PM, Satnam Singh <satnam@google.com> wrote:

> I wrote the wrong thing at first, I updated it to: labels.Set{"name":
> name}.AsSelector()
>
> S
>
> On Tue, Jan 13, 2015 at 1:45 PM, Brendan Burns <notifications@github.com>
> wrote:
>
>> In pkg/kubectl/run.go
>> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3424#discussion-diff-22897598>
>> :
>>
>> > +		{"labels", false},
>> > +		{"name", true},
>> > +		{"replicas", true},
>> > +		{"image", true},
>> > +		{"command", false},
>> > +	}
>> > +}
>> > +
>> > +func (BasicReplicationController) Generate(params map[string]string) (runtime.Object, error) {
>> > +	labelString, found := params["labels"]
>> > +	var labels map[string]string
>> > +	if found && len(labelString) > 0 {
>> > +		labels = ParseLabels(labelString)
>> > +	} else {
>> > +		labels = map[string]string{
>> > +			"name": params["name"],
>>
>> Selector interface won't work when I try to put it back into the api
>> object (Selector is not map[string]string)
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3424/files#r22897598>
>> .
>>
>
>

bgrant0607 replied:
I think creating a label from the name by default is useful, esp. once we support uniquified names, bulk creation, and cloning. However, one issue with calling it `name` is that it won't precisely match the object names in those cases, nor of the pods created by the replication controller. This is why I lean towards something less generic, such as `group`.

Ideally, all labels we add would be namespaced, so as not to conflict with user-provided labels -- we use `kubectl.kubernetes.io/` elsewhere. I understand that the usability argument against that, though. One compromise would be to use the un-namespaced key if the user doesn't specify any labels and the fully qualified, namespaced key if they do.

bgrant0607 replied:
My last comment here was not addressed in the last push. `name` generates confusion with respect to the metadata "name" field in the API. I'm flexible about what we replace it with: group, set, uservice, app, deployment, run-name, ...

smarterclayton replied:
App :).  


> On Jan 13, 2015, at 5:59 PM, bgrant0607 <notifications@github.com> wrote:
> 
> In pkg/kubectl/run.go:
> 
> > +		{"labels", false},
> > +		{"name", true},
> > +		{"replicas", true},
> > +		{"image", true},
> > +		{"command", false},
> > +	}
> > +}
> > +
> > +func (BasicReplicationController) Generate(params map[string]string) (runtime.Object, error) {
> > +	labelString, found := params["labels"]
> > +	var labels map[string]string
> > +	if found && len(labelString) > 0 {
> > +		labels = ParseLabels(labelString)
> > +	} else {
> > +		labels = map[string]string{
> > +			"name": params["name"],
> My last comment here was not addressed in the last push. name generates confusion with respect to the metadata "name" field in the API. I'm flexible about what we replace it with: group, set, uservice, app, deployment, run-name, ...
> 
> —
> Reply to this email directly or view it on GitHub.
>

brendandburns replied:
sorry I missed that in the earlier comment.  Switched to ```run-container``` as the label

brendandburns commented:
Updated cli.md (I updated the code, but not the doc, we should really auto-generate :P)

bgrant0607 commented on pkg/kubectl/cmd/helpers.go:
```diff
@@ -171,3 +173,17 @@ func ReadConfigDataFromLocation(location string) ([]byte, error) {
+
+func PrintObject(cmd *cobra.Command, obj runtime.Object, f *Factory, out io.Writer) {
+	mapper, _ := f.Object(cmd)
+	defaultVersion, kind, err := mapper.VersionAndKindForResource("replicationcontrollers")
```
Is there not an easy way to make this work for any kind? Could it be extracted from the object? Or perhaps just make kind a parameter to the function? 

This is how I did it in the apiserver:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/apiserver/apiserver.go#L109

Maybe @smarterclayton knows a better way.

bgrant0607 commented on docs/cli.md:
```diff
@@ -125,6 +125,65 @@ Usage:
+  $ kubectl run-container nginx --image=dockerfile/nginx
+  <starts a single instance of nginx>
+
+  $ kubectl run-container nginx --image=dockerfile/nginx --command="nginx run" --count=5
```
replicas

brendandburns replied:
done.

bgrant0607 commented on pkg/kubectl/cmd/helpers.go:
```diff
@@ -26,6 +27,7 @@ import (
 	"strings"
 	"time"
 
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
```
This file should be reverted.

brendandburns replied:
done.

bgrant0607 commented on pkg/kubectl/cmd/printing.go:
```diff
@@ -0,0 +1,85 @@
+// PrintObject prints an api object given command line flags to modify the output format
+func PrintObject(cmd *cobra.Command, obj runtime.Object, f *Factory, out io.Writer) {
+	mapper, _ := f.Object(cmd)
+	defaultVersion, kind, err := mapper.VersionAndKindForResource("replicationcontrollers")
```
See my comment on the previous version.

brendandburns replied:
done.

bgrant0607 commented on pkg/kubectl/cmd/run.go:
```diff
@@ -0,0 +1,84 @@
+  $ kubectl run-container nginx --image=dockerfile/nginx
+  <starts a single instance of nginx>
+
+  $ kubectl run-container nginx --image=dockerfile/nginx --command="nginx run" --count=5
```
replicas

brendandburns replied:
done.

bgrant0607 commented on pkg/kubectl/cmd/run.go:
```diff
@@ -0,0 +1,84 @@
+		Use:   "run-container <name> --image=<image> [--command=<some command here>] [--replicas=replicas] [--dry-run=<bool>]",
+		Short: "Run a particular image on the cluster.",
+		Long: `Create and run a particular image, possibly replicated.
+Creates a replication controller to manage the created image(s)
```
"the created containers(s)", to avoid confusion if/when/where there is a way to create images (e.g., in Openshift)

brendandburns replied:
done.

bgrant0607 commented on pkg/kubectl/generate.go:
```diff
@@ -0,0 +1,69 @@
+// Generators is a global list of known generators.
+// TODO: Dynamically create this from a list of template files?
+var Generators map[string]Generator = map[string]Generator{
+	"run-container-controller-v1": BasicReplicationController{},
```
I would remove "controller" from the generator name, because:
1. it would more closely match the command name, thus more clearly establishing a pattern
2. it would reduce confusion about whether "run-container" is active in the same way a real API object is -- it's not, it's a macro expansion, which in this case just happens to expand into a controller
3. we're going to rename ReplicationController O(tomorrow)

I'd also like to use `/` as the separator between generator name and version, in case we want to reason about generator version in a more first-class manner: `run-container/v1`.

brendandburns replied:
done.

bgrant0607 commented on pkg/kubectl/generate.go:
```diff
@@ -0,0 +1,69 @@
+
+// GeneratorParam is a parameter for a generator
+type GeneratorParam struct {
+	Name     string
```
Add a TODO: facilitate structured json generator input schemas.

This approach would force generators that accepted nested or repeated entities to flatten their inputs rather than just accept json, which is my preferred approach.

However, since this isn't exposed to users, it isn't critical to fix in this PR.

brendandburns replied:
done.

bgrant0607 commented on pkg/kubectl/run.go:
```diff
@@ -0,0 +1,100 @@
+}
+
+func (BasicReplicationController) Generate(params map[string]string) (runtime.Object, error) {
+	labelString, found := params["labels"]
```
// TODO: extract this flag to a central location, when such a location exists.

I really want custom support for some curated labels, in which case we'll really want a centralized implementation of construction of the labels map.

brendandburns replied:
done.

bgrant0607 commented on pkg/kubectl/run.go:
```diff
@@ -0,0 +1,100 @@
+			Replicas: count,
+			Selector: labels,
+			Template: &api.PodTemplateSpec{
+				Spec: api.PodSpec{
```
You need to set the labels in the PodTemplateSpec.

brendandburns replied:
good catch.  done.

bgrant0607 commented on pkg/kubectl/run.go:
```diff
@@ -0,0 +1,100 @@
+	if len(cmd) == 0 {
+		return nil
+	}
+	return strings.Split(cmd, " ")
```
This seems totally dangerous. What if there are quoted or escaped spaces in the command? We need a command string in the API, as per #2316. Could we just not expose this until we have that? It seems likely to lead to user confusion.

brendandburns replied:
removed.

bgrant0607 commented on pkg/kubectl/run.go:
```diff
@@ -0,0 +1,100 @@
+	return &controller, nil
+}
+
+func ParseLabels(labelString string) map[string]string {
```
Also TODO hoisting this to a common location. I'd like create to support this.

brendandburns replied:
done.

bgrant0607 commented:
Finished pass over the PR. Please review my outdated comments for ones that weren't addressed.

brendandburns commented:
Comments addressed (I believe I got them all before I pushed)

bgrant0607 commented on docs/cli.md:
```diff
@@ -125,6 +125,64 @@ Usage:
+  <just print the corresponding API objects, don't actually send them to the apiserver>
+
+Usage: 
+  kubectl run-container <name> --image=<image> [--command=<some command here>] [--replicas=replicas] [--dry-run=<bool>] [flags]
```
Thanks for removing command. Now we need to update the doc. (Yes, auto-generation would be cool.)

brendandburns replied:
done.

bgrant0607 commented on pkg/kubectl/cmd/run.go:
```diff
@@ -0,0 +1,83 @@
+
+func (f *Factory) NewCmdRunContainer(out io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "run-container <name> --image=<image> [--command=<some command here>] [--replicas=replicas] [--dry-run=<bool>]",
```
"command" also needs to be removed here.

brendandburns replied:
done.

bgrant0607 commented:
Thanks for the changes. Getting close.

bgrant0607 commented:
The Travis failures are real.

smarterclayton commented:
The changes I landed as a consequence of get and the context change affect you (I altered those methods).  There's an outstanding fix I need to make to default api-version to the kube version and ensure non-uniform api objects can all get converted to whatever api version is closest to what the user asks

brendandburns commented:
Comments addressed, I'm specifying a constant in place of the ```--api-version``` flag since it seems to no longer be present anywhere except for in some tests.

bgrant0607 commented:
LGTM, but apparently there are merge conflicts now.

brendandburns commented:
Squashed and rebased.

Thanks!
--brendan

bgrant0607 commented:
Will merge in the morning.

smarterclayton commented:
APIVersion will be back in some form, but we should be all means stop directly accessing the flag.


> On Jan 13, 2015, at 8:03 PM, Brendan Burns <notifications@github.com> wrote:
> 
> Comments addressed, I'm specifying a constant in place of the --api-version flag since it seems to no longer be present anywhere except for in some tests.
> 
> —
> Reply to this email directly or view it on GitHub.
>

brendandburns commented:
Yeah, actually when I rebased, mapping.APIVersion was in place, so I
removed the hard coding in favor of mapping.APIVersion

--brendan

On Wed, Jan 14, 2015 at 3:34 PM, Clayton Coleman <notifications@github.com>
wrote:

> APIVersion will be back in some form, but we should be all means stop
> directly accessing the flag.
>
>
> > On Jan 13, 2015, at 8:03 PM, Brendan Burns <notifications@github.com>
> wrote:
> >
> > Comments addressed, I'm specifying a constant in place of the
> --api-version flag since it seems to no longer be present anywhere except
> for in some tests.
> >
> > —
> > Reply to this email directly or view it on GitHub.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3424#issuecomment-70014485>
> .
>
